src/reducers/sources.js

```diff
@@ -149,40 +149,40 @@ function update(
   return state;
 }
 
-// TODO: Action is coerced to `any` unfortunately because how we type
-// asynchronous actions is wrong. The `value` may be null for the
-// "start" and "error" states but we don't type it like that. We need
-// to rethink how we type async actions.
-function _updateText(state, action: any): Record<SourcesState> {
+function getTextFromAction(action: any) {
   const source = action.source;
   const sourceText = action.value;
 
   if (action.status === "start") {
-    // Merge this in, don't set it. That way the previous value is
-    // still stored here, and we can retrieve it if whatever we're
-    // doing fails.
-    return state.mergeIn(["sourcesText", source.id], {
-      loading: true
-    });
+    return { loading: true };
+  } else if (action.status === "error") {
+    return { error: action.error };
   }
+  return {
+    text: sourceText.text,
+    id: source.id,
+    contentType: sourceText.contentType,
+    loading: false
+  };
+}
 
-  if (action.status === "error") {
-    return state.setIn(
-      ["sourcesText", source.id],
-      I.Map({
-        error: action.error
-      })
-    );
+// TODO: Action is coerced to `any` unfortunately because how we type
+// asynchronous actions is wrong. The `value` may be null for the
+// "start" and "error" states but we don't type it like that. We need
+// to rethink how we type async actions.
+function updateText(state, action: any): Record<SourcesState> {
+  const source = action.source;
+  const text = getTextFromAction(action);
+  const updatedState = state.setIn(["sourcesText", source.id], I.Map(text));
+  return updateSource(updatedState, text);
+}
+
+function updateSource(state: State, source: Object | Source) {
+  if (!source.id) {
+    return state;
   }
 
-  return state.setIn(
-    ["sourcesText", source.id],
-    I.Map({
-      text: sourceText.text,
-      id: source.id,
-      contentType: sourceText.contentType
-    })
-  );
+  return state.mergeIn(["sources", source.id], source);
```

Comment:
maybe we can have a comment here about why we are merging, rather than setting? I saw above there was such a comment in the previous version of the code, might be helpful here too